sbx/sac/sac.py

```diff
@@ -331,62 +327,112 @@ def temperature_loss(temp_params):
         return ent_coef_state, ent_coef_loss
 
     @classmethod
-    @partial(jax.jit, static_argnames=["cls", "gradient_steps"])
+    def update_actor_and_temperature(
+        cls,
+        actor_state: RLTrainState,
+        qf_state: RLTrainState,
+        ent_coef_state: TrainState,
+        observations: jax.Array,
+        target_entropy: ArrayLike,
+        key: jax.random.KeyArray,
+    ):
+        (actor_state, qf_state, actor_loss_value, key, entropy) = cls.update_actor(
+            actor_state,
+            qf_state,
+            ent_coef_state,
+            observations,
+            key,
+        )
+        ent_coef_state, ent_coef_loss_value = cls.update_temperature(target_entropy, ent_coef_state, entropy)
+        return actor_state, qf_state, ent_coef_state, actor_loss_value, ent_coef_loss_value, key
+
+    @classmethod
+    @partial(jax.jit, static_argnames=["cls", "gradient_steps", "policy_delay_interval", "policy_delay_offset"])
     def _train(
         cls,
         gamma: float,
         tau: float,
-        target_entropy: np.ndarray,
+        target_entropy: ArrayLike,
         gradient_steps: int,
         data: ReplayBufferSamplesNp,
-        policy_delay_indices: flax.core.FrozenDict,
+        policy_delay_interval: int,
+        policy_delay_offset: int,
         qf_state: RLTrainState,
         actor_state: TrainState,
         ent_coef_state: TrainState,
-        key,
+        key: jax.random.KeyArray,
     ):
-        actor_loss_value = jnp.array(0)
-
-        for i in range(gradient_steps):
-
-            def slice(x, step=i):
-                assert x.shape[0] % gradient_steps == 0
-                batch_size = x.shape[0] // gradient_steps
-                return x[batch_size * step : batch_size * (step + 1)]
-
+        assert data.observations.shape[0] % gradient_steps == 0
+        batch_size = data.observations.shape[0] // gradient_steps
+
+        carry = {
+            "actor_state": actor_state,
+            "qf_state": qf_state,
+            "ent_coef_state": ent_coef_state,
+            "key": key,
+            "info": {
+                "actor_loss": jnp.array(0.0),
+                "qf_loss": jnp.array(0.0),
+                "ent_coef_loss": jnp.array(0.0),
+            },
+        }
+
+        def one_update(i: int, carry: Dict[str, Any]) -> Dict[str, Any]:
+            actor_state = carry["actor_state"]
+            qf_state = carry["qf_state"]
+            ent_coef_state = carry["ent_coef_state"]
+            key = carry["key"]
+            info = carry["info"]
+            batch_obs = jax.lax.dynamic_slice_in_dim(data.observations, i, batch_size)
```

Comment:
`start_index` should be `batch_size * i`, no? (instead of `i` here)

Reply:
Nice catch. I fixed it in [4cd5f27](https://github.com/araffin/sbx/pull/21/commits/4cd5f27e5b6c297328499b7dcb8fbbcaa0f0e730).